File: backend/app/api/v1/batches.py
"""Endpoints de Carga Masiva (Batch) — procesamiento asíncrono."""
import uuid

from fastapi import APIRouter, File, Form, UploadFile

from app.api.deps import CurrentUserAny, DbSession
from app.schemas.batch import BatchDocumentRead, BatchStatusRead, BatchSubmitResponse
from app.services.batch_service import BatchService
from app.services.document_service import DocumentService

router = APIRouter(prefix="/batches", tags=["batches"])


@router.post("", response_model=BatchSubmitResponse, status_code=202)
def submit_batch(
    db: DbSession,
    user: CurrentUserAny,
    file: UploadFile = File(...),
    processing_mode: str = Form("almacenado"),
    template_id: str = Form(""),
):
    tid = uuid.UUID(template_id) if template_id else None
    batch = BatchService(db).submit_batch(user, file, processing_mode, template_id=tid)
    return BatchSubmitResponse(
        batch_id=batch.id,
        status=batch.status,
        total_files=batch.total_files,
        message=f"Lote recibido con {batch.total_files} archivo(s). Se procesará en segundo plano y recibirá una notificación al finalizar.",
    )


@router.get("", response_model=list[BatchStatusRead])
def list_batches(db: DbSession, user: CurrentUserAny, limit: int = 20):
    return BatchService(db).list_batches(user, limit)


@router.get("/{batch_id}", response_model=BatchStatusRead)
def get_batch_status(batch_id: uuid.UUID, db: DbSession, user: CurrentUserAny):
    return BatchService(db).get_batch(user, batch_id)


@router.get("/{batch_id}/documents", response_model=list[BatchDocumentRead])
def list_batch_documents(batch_id: uuid.UUID, db: DbSession, user: CurrentUserAny):
    batch = BatchService(db).get_batch(user, batch_id)
    return [
        BatchDocumentRead(
            id=doc.id,
            original_filename=doc.original_filename,
            status=doc.status,
            document_type=doc.document_type,
            page_count=doc.page_count,
        )
        for doc in batch.documents
    ]

File: frontend/src/api/batches.ts
import { apiClient } from "./client";

export interface BatchSubmitResponse {
  batch_id: string;
  status: string;
  total_files: number;
  message: string;
}

export interface BatchStatusRead {
  id: string;
  status: string;
  processing_mode: string;
  total_files: number;
  processed_files: number;
  failed_files: number;
  created_at: string;
  completed_at: string | null;
}

export interface BatchDocumentRead {
  id: string;
  original_filename: string;
  status: string;
  document_type: string | null;
  page_count: number;
}

export async function submitBatch(
  zipFile: File,
  processingMode: "express" | "almacenado",
  templateId?: string
): Promise<BatchSubmitResponse> {
  const formData = new FormData();
  formData.append("file", zipFile);
  formData.append("processing_mode", processingMode);
  if (templateId) formData.append("template_id", templateId);
  const { data } = await apiClient.post("/batches", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

export async function listBatches(limit = 20): Promise<BatchStatusRead[]> {
  const { data } = await apiClient.get("/batches", { params: { limit } });
  return data;
}

export async function getBatchStatus(batchId: string): Promise<BatchStatusRead> {
  const { data } = await apiClient.get(`/batches/${batchId}`);
  return data;
}

export async function listBatchDocuments(batchId: string): Promise<BatchDocumentRead[]> {
  const { data } = await apiClient.get(`/batches/${batchId}/documents`);
  return data;
}
